packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js

```diff
@@ -207,18 +207,49 @@ RocketChat.callbacks.add('afterSaveMessage', function(message, room, userId) {
 		userIds.push(s.u._id);
 	});
 	const userSettings = {};
-	RocketChat.models.Users.findUsersByIds(userIds, { fields: { 'settings.preferences.audioNotifications': 1, 'settings.preferences.desktopNotifications': 1, 'settings.preferences.mobileNotifications': 1 } }).forEach((user) => {
+	RocketChat.models.Users.findUsersByIds(userIds, { fields: { 'settings.preferences.audioNotifications': 1, 'settings.preferences.desktopNotifications': 1, 'settings.preferences.mobileNotifications': 1, 'settings.preferences.snoozeNotifications': 1, 'settings.preferences.doNotDisturb': 1 } }).forEach((user) => {
 		userSettings[user._id] = user.settings;
 	});
 
 	subscriptions.forEach(subscription => {
-		if (subscription.disableNotifications) {
+		const preferences = userSettings[subscription.u._id] ? userSettings[subscription.u._id].preferences || {} : {};
+
+		let snoozeNotifications = !!(subscription.snoozeNotifications && subscription.snoozeNotifications.duration && subscription.snoozeNotifications.finalDateTime && moment().isBefore(subscription.snoozeNotifications.finalDateTime));
+
+		if (!snoozeNotifications) {
+			snoozeNotifications = !!(preferences.snoozeNotifications && preferences.snoozeNotifications.duration && preferences.snoozeNotifications.finalDateTime && moment().isBefore(preferences.snoozeNotifications.finalDateTime));
+		}
+
+		let doNotDisturb = false;
+
+		const doNotDisturbValidate = (doNotDisturbObject) => {
+			if (doNotDisturbObject.repeatFor && (doNotDisturbObject.repeatFor === 'every day' || (doNotDisturbObject.limitDateTime && moment().isBefore(doNotDisturbObject.limitDateTime)))) {
+				const initialMoment = moment(doNotDisturbObject.initialTime, 'HH:mm');
```

Comment:
What happens if server's time (GMT) is different from user's time?

Reply:
Applied user's UTC offset on 'do not disturb' check at server side :-)